src/hooks/cgroups/pbs_cgroups.PY

```diff
@@ -308,14 +352,34 @@ def decode_dict(data):
     """
     ret = {}
     for key, value in list(data.items()):
-        if isinstance(key, (bytes, bytearray)):
-            key = str(key, 'utf-8')
-        if isinstance(value, (bytes, bytearray)):
-            value = str(value, 'utf-8')
+        # first the key
+        if isinstance(key, str):
+            pass
+        elif (sys.version_info[0] < 3) and isinstance(key, unicode):
+            key = key.encode('utf-8')
+        elif isinstance(key, (bytes, bytearray)):
+            if sys.version_info[0] >= 3:
+                key = str(key, 'utf-8')
+            else:
+                key = str(key)
+
+        # now the value
+        if isinstance(value, str):
+            pass
+        elif (sys.version_info[0] < 3) and isinstance(value, unicode):
+            value = value.encode('utf-8')
+        elif isinstance(value, (bytes, bytearray)):
+            if sys.version_info[0] >= 3:
+                value = str(value, 'utf-8')
+            else:
+                key = str(value)
```

Comment:
With all respect, I think this is out of scope. Refactoring the code in such a way is risky, and I feel this should be done by the original author who broke "master" for Python 2. Here, there are very localized changes working around bugs that make the hook broken on Python2 systems, to minimize risks.

A reusable function may not even be the best solution. Perhaps it would be best to use a string-like type derived from either bytes or unicode strings on both Python2 and Python3 and to use it throughout the code, but there is a routine where systemd converts strings into its own peculiar flavour of escape sequences that might actually act differently for these two base types (suggesting we'd better use "bytes" rather than Unicode).

Reply:
I'd like to point out that we no longer support python2.  Making this hook support python 2 might be a good thing, but it is not something PBSPro master cares about.  Master only supports python 3.  If you want to make this hook support python 2 and 3, I'd suggest doing it the right way.

Reply:
"I'd like to point out that we no longer support python2."

I'd like to point out that no currently released version and generally available release of PBSPro is using python3. Current customers would rather have access to a hook that works for them now, rather than in a distant cornucopia in which even PBSPro 19.x has been retired.

Indeed, once sites do convert to a PBSPro version that does support Python 3, many of those sites will still have other clusters still running an older version of PBSPro, and for them it is important for the cgroup hook to have consistent behaviour across the entire complex of different clusters.

"Making this hook support python 2 might be a good thing, but it is not something PBSPro master cares about."

This is the cgroup hook, which, as a hook, has utility well beyond its use in a particular release of PBSPro.

Sites have been using the cgroup hook in one of its incarnations in PBSPro 13.0.401, 14.x, 18.x and 19.x. Of the more than ten sites I know, exactly zero sites have been able to use the hook as shipped in their production environment for long. Sometimes due to bugs in the hook, sometimes due to changes in the operating system behaviour that were unanticipated (especially with respect to systemd), sometimes because of interoperability with other hooks or foreign software (including the Docker hook and Nvidia software for GPUs).

All these customers relied on backward compatibility of more recent versions of the hook to save them, since very often the issues were addressed in more recent hooks.

The "master" hook was not created in a vacuum: it was derived from 18.x/19.x hooks that were Python2 compatible but not Python3 compatible. Unfortunately, the author that made the hook Python3 compatible did not take this real life need for backward compatibility into account and morphed code that worked _only_ in Python2 into code that worked _only_ in Python3.

In doing that, the hook as rewritten would not even be useful should anyone decide to e.g. release 19.2.6 and want to fix all the issues that were being addressed in the master hook.

"doing it the right way"

Unfortunately, there _is_ no right way. The Python maintainers have _decided_ to break backward compatibility of Python in this respect. There are two underlying "string like classes", "bytes" (which hold only 8-byte characters and have to resort to an encoding, usually UTF-8 for Linux hooks, to contain some characters) and unicode strings (which contains unicode characters that may happen to be 8-byte characters in most cases. 

In Python 2, strings are essentially "bytes" by another name. In Python 3, strings are now unicode strings, and the maintainers have decided to even _remove_ the keywords that would let you use the Python 2 syntax for unicode strings because they felt it was redundant and confusing. That means strings, which are returned by many, many methods from many, many classes, are now fundamentally different, with different methods that exist for the classes, with encoding and decoding flipped with respect to "Python 2 strings". That makes it well nigh impossible to make code that works _without_ checking for the Python version.

As to why I chose not to refactor: given the last thing we want is for customers to run into some TypeErrors or methods not found in class X that simply crash the hook, I took the approach with minimal risk, which is to re-borrow the existing (and tested!) 18.x code snippets and select whether to run the "old" or "new" snippets of code depending on the Python version, to make the hook look as close as possible to the tested 18.x hook when on Python 2 and as close as possible to the existing master when on Python 3.

Refactoring this to have these Python version checks in less than the 6 routines they are in now isn't that easy. Firstly, all of these routines do slightly different string operations, with slightly different Python 2 vs Python 3 syntax, so just making a single routine is not easy. Changing classes by using a self-defined class that is consistent in behaviour on both Python 2 and Python 3 would seem to offer a solution, but whatever we pick, it will not simply combine with "plain strings" on _one_ of the two Python versions (in fact, what the current code does is isolate the differences between Python 2 and Python 3 in a very small number of routines, with the rest of the code "naively" using just plain strings).

If we _completely_ change the Python 2 version to handle unicode strings everywhere, then it would be easy to indeed make small changes to get a Python 3 version. That is not a trivial effort -- in fact, Python 3 broke backward compatibility because it was harder than necessary in Python 2 (given the fact that plain strings vs. unicode strings need encdoe/decode to convert to eahc other, and that plain strings are returned by many classes).

Reply:
I understand what you are saying, but that is not how PBS versioning works.  The version in master is valid for the next release.  It is not meant to be used in a previous release of PBS.  If a bug fix is added into master and it is needed in a previous release, we add it to the patch branch for that release.  It will come up in the next patch (e.g. the security fix in 19.1.3).  Taking a hook that is meant for a future release and using it on a previous release is not necessarily what we want to happen.  We won't test that.  This hook will be tested with 2020.1 when it comes out.  It will not be tested with 19 or 18 or 13.  Using code in this way is not supported.  Think about when suspend/resume is supported in the hook?  These are hook events that do not exist in 19.

Reply:
Understood. I am _NOT_ advocating supporting backward compatibility forever, BTW. It may well be that some day we'll want to implement new features in the hook that are a lot easier to do just in Python 3. But the changes introduced recently were not that, they were code that was originally Python 2 code translated to Python 3-only code.

It's one thing to say you won't support backward compatibility ad infinitum, and quite another to almost actively implement hurdles for it.